### PR TITLE
Spark documentation

### PR DESCRIPTION
**Pull request by Sami1309**
# Description

Docs for spark implementation

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: docs/providers/spark.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+Using the client, a user can define transformtions using dataframes, which transform into new tables. The function name and decorator variant label can be used as an identifier to further chain transformations, and use the new tables to create training sets.
+
+Current support requires an [AWS EMR](https://aws.amazon.com/emr/) cluster and an [S3](https://aws.amazon.com/s3/) bucket to store the featureform generated tables, feature definitions and training sets.
+Features can be copied from spark to an inference store (ex: Redis)(link to page) for real-time feature serving
+
+Data is stored in parquet tables in the bucket. The spark store executes a spark job on your cluster to transform the data and copy it to its relevant sources for training and serving
+
+#### Requirements
+	Amazon AWS account with:
```
Comment (sdreyer) [inline docs/providers/spark.md:16]: Make these bullet points

File: docs/providers/spark.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Spark
+
+Featureform supports [Spark on AWS](https://aws.amazon.com/emr/features/spark/) as an Offline Store
+
+## Implementation <a href="#implementation" id="implementation"></a>
+Featureform implements spark as a transformation layer. The results of the transformations and table creations are stored in S3.
+
+Using the client, a user can define transformtions using dataframes, which transform into new tables. The function name and decorator variant label can be used as an identifier to further chain transformations, and use the new tables to create training sets.
```
Comment (sdreyer) [inline docs/providers/spark.md:8]: "transformtions"

Comment (sdreyer) [reply docs/providers/spark.md:8]: dataframes and sql

File: docs/providers/spark.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Current support requires an [AWS EMR](https://aws.amazon.com/emr/) cluster and an [S3](https://aws.amazon.com/s3/) bucket to store the featureform generated tables, feature definitions and training sets.
+Features can be copied from spark to an inference store (ex: Redis)(link to page) for real-time feature serving
```
Comment (sdreyer) [inline docs/providers/spark.md:11]: Missing link for redis

File: docs/providers/spark.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+Data is stored in parquet tables in the bucket. The spark store executes a spark job on your cluster to transform the data and copy it to its relevant sources for training and serving
```
Comment (sdreyer) [inline docs/providers/spark.md:13]: Confused about what you mean by "copy it to its relevant sources for training and serving"

File: docs/providers/spark.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Requirements
+	Amazon AWS account with:
+		AWS S3 Bucket
+		AWS EMR Cluster
+			Runs Spark version >2
+
+### Transformation Sources
+
+SQL transformations are used to create a view. By default, those views are materialized and updated according to the schedule parameter. Deprecated transformations are converted to un-materialized views to save storage space.
```
Comment (sdreyer) [inline docs/providers/spark.md:23]: Is this SQL and dataframes?

File: docs/providers/spark.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    emr_cluster_region = "us-east-1",
+    bucket_region = "us-east-2",
+    aws_access_key_id = "<access-key-id>",
+    aws_secret_access_key = "<aws-secret-access-key>",):
+```
+{% endcode %}
+
+### Dataframe Transformations
+While you can also create SQL transformations as in the other offline stores like [Redis](providers/redis.md), Spark uniquely allows you to perform dataframe transformations using your source tables as inputs.
+with your registered spark store, you can perform
```
Comment (sdreyer) [inline docs/providers/spark.md:57]: capitalize With

File: docs/providers/spark.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    return df
+
+@spark.sql_transformation()        # Sources are added as inputs
+def max_transaction_amount():                           # Sources can be manipulated by adding them as params
+    """the average transaction amount for a user """
+    return "SELECT CustomerID as user_id, max(TransactionAmount) " \
+           "as max_transaction_amt from {{transactions.kaggle}} GROUP BY user_id"
+```
+{% endcode %}
+and use the result of your transformation as the input to another one
```
Comment (sdreyer) [inline docs/providers/spark.md:73]: Period

File: docs/providers/spark.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ff.register_snowflake(
+    name = "spark_offline_store"
+    description = "A spark provider that can create transformations and training sets",
+    team = "featureform data team",
+    emr_cluster_id = "j-ExampleCluster",
+    bucket_path = "example-bucket-path", //excluding the "S3://" prefix
+    emr_cluster_region = "us-east-1",
+    bucket_region = "us-east-2",
+    aws_access_key_id = "<access-key-id>",
+    aws_secret_access_key = "<aws-secret-access-key>",):
```
Comment (sdreyer) [inline docs/providers/spark.md:51]: newline the )

File: docs/providers/spark.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    return "SELECT CustomerID as user_id, max(TransactionAmount) " \
+           "as max_transaction_amt from {{transactions.kaggle}} GROUP BY user_id"
+```
+{% endcode %}
+and use the result of your transformation as the input to another one
+{% code title="dataframe_registration.py" %}
+```python
+@spark.sql_transformation()
+def average_user_transaction():                           # Sources can be manipulated by adding them as params
+    "SELECT * FROM {{average_user_transaction.default}} INNER JOIN {{max_transaction_amount}} ON " \
```
Comment (sdreyer) [inline docs/providers/spark.md:78]: This needs a return

File: provider/spark.go [hunk 10]
```diff
@@ -381,7 +381,7 @@ func (e *EMRExecutor) RunSparkJob(args []string) error {
 					Jar:  aws.String("command-runner.jar"), //jar file for running pyspark scripts
 					Args: args,
 				},
-				ActionOnFailure: emrTypes.ActionOnFailureCancelAndWait,
+				ActionOnFailure: emrTypes.ActionOnFailureContinue,
```
Comment (sdreyer) [inline provider/spark.go:384]: Whats this do?

File: docs/providers/spark.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    name = "spark_offline_store"
+    description = "A spark provider that can create transformations and training sets",
+    team = "featureform data team",
+    emr_cluster_id = "j-ExampleCluster",
+    bucket_path = "example-bucket-path", #excluding the "S3://" prefix
+    emr_cluster_region = "us-east-1",
+    bucket_region = "us-east-2",
+    aws_access_key_id = "<access-key-id>",
+    aws_secret_access_key = "<aws-secret-access-key>",
+    ):
```
Comment (sdreyer) [inline docs/providers/spark.md:46]: There shouldnt be a colon here

File: docs/providers/spark.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+### Dataframe Transformations
+Using Spark with Featureform, a user can define transformations in SQL like with other offline providers.
+
+{% code title="sql_transformation.py" %}
+```python
+@spark.sql_transformation()
+def max_transaction_amount():
+    """the average transaction amount for a user """
+    return "SELECT CustomerID as user_id, " \
+    "max(TransactionAmount) " "as max_transaction_amt " \
```
Comment (sdreyer) [inline docs/providers/spark.md:59]: This query has extra quotes in it. It was fine when it was 2 lines

File: docs/providers/spark.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+In addition, registering a provider via Spark allows you to perform DataFrame transformations using your source tables as inputs.
+
+{% code title="dataframe_transformation.py" %}
+```python
+@spark.df_transformation(
+    inputs=[("transactions", "kaggle")], 
+    variant="default")
+def average_user_transaction(df):
+    from pyspark.sql.functions import avg
+        df.groupBy("CustomerID")
```
Comment (sdreyer) [inline docs/providers/spark.md:73]: Think this df one was also fine as one line. Typically want to newline things if they break 100 characters. Start to get weird when you start bunching everything vertically

File: docs/providers/spark.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+Your [AWS access key id and AWS secret access key](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html) are used as credentials when registering your Spark Offline Store.
+
+Your EMR cluster must be running and support [Spark](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark.html).
+
+{% code title="spark_config.py" %}
+```python
+import featureform as ff
+
+ff.register_snowflake(
```
Comment (ahmadnazeri) [inline docs/providers/spark.md:36]: register_spark?

**Review by sdreyer — DISMISSED** [review]

**Review by sdreyer — DISMISSED** [review]

File: docs/providers/spark.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    emr_cluster_id = "j-ExampleCluster",
+    bucket_path = "example-bucket-path", #excluding the "S3://" prefix
+    emr_cluster_region = "us-east-1",
+    bucket_region = "us-east-2",
+    aws_access_key_id = "<access-key-id>",
+    aws_secret_access_key = "<aws-secret-access-key>",
+    )
+```
+{% endcode %}
+
```
Comment (ahmadnazeri) [inline docs/providers/spark.md:49]: include `spark.register_parquet_file`

**Review by sdreyer — APPROVED** [review]

**Pull request merged**